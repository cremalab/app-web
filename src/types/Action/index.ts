import { Action as ReduxAction } from "redux"

/**
 * `Action`
 * ---------------------------------------------
 * Extends redux `Action` adding `payload`
 */

export interface Action<Type extends PropertyKey, Payload>
  extends ReduxAction<Type> {
  payload: Payload
}

/**
 * `Action`
 * ------------------
 * Returns a `Tagged<Tag, Data>`
 **/

export const Action = <Type extends PropertyKey, Payload>(
  type: Type,
  payload: Payload,
): Action<Type, Payload> => ({
  type,
  payload,
})
