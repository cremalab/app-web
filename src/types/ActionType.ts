export enum ActionType {
  TodoAdd = "TodoAdd",
  TodoRemove = "TodoRemove",
  TodoUpdate = "TodoUpdate",
}
