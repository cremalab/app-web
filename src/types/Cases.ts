import { Tagged } from "./Tagged"

export type Cases<
  A extends Tagged<Tag, unknown>,
  Return,
  Tag extends PropertyKey = PropertyKey
> = {
  [K in A["tag"]]: A extends Tagged<K, infer Data> ? (x: Data) => Return : never
}
