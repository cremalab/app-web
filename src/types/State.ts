import { EntityTodo } from "./EntityTodo"
import { StateCollection } from "./StateCollection"

export interface State {
  todos: StateCollection<EntityTodo>
}
