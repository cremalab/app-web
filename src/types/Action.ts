import { Tagged } from "./Tagged"
import { ActionType } from "./ActionType"

export interface Action<Payload = null, Type extends PropertyKey = ActionType>
  extends Tagged<Type> {
  payload: Payload
}
