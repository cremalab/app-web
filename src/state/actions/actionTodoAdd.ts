import { ActionType } from "../../types/ActionType"
import { makeActionCreator } from "../../utils/makeActionCreator"
import { EntityTodo } from "../../types/EntityTodo"
import { TaggedActionOfActionCreator } from "../../types/TaggedActionOfActionCreator"

export const actionTodoAdd = makeActionCreator(
  ActionType.TodoAdd,
  (a: EntityTodo) => a,
)

export type TaggedActionTodoAdd = TaggedActionOfActionCreator<
  typeof actionTodoAdd
>
