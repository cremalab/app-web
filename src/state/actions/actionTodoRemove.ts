import { ActionType } from "../../types/ActionType"
import { makeActionCreator } from "../../utils/makeActionCreator"
import { EntityTodo } from "../../types/EntityTodo"

export const actionTodoRemove = makeActionCreator(
  ActionType.TodoRemove,
  (a: EntityTodo["id"]) => a,
)

export type ActionTodoRemove = ReturnType<typeof actionTodoRemove>
