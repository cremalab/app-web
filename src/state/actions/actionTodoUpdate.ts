import { ActionType } from "../../types/ActionType"
import { makeActionCreator } from "../../utils/makeActionCreator"
import { EntityTodo } from "../../types/EntityTodo"
import { Update } from "../../types/Update"
import { TaggedActionOfActionCreator } from "../../types/TaggedActionOfActionCreator"

export const actionTodoUpdate = makeActionCreator(
  ActionType.TodoUpdate,
  (
    id: Update<EntityTodo>["id"],
    properties: Update<EntityTodo>["properties"],
  ): Update<EntityTodo> => ({
    id,
    properties,
  }),
)

export type TaggedActionTodoUpdate = TaggedActionOfActionCreator<
  typeof actionTodoUpdate
>
