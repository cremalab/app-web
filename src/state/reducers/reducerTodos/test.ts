import { actionTodoAdd } from "../../actions/actionTodoAdd"
import { reducerTodos } from "."
import { actionTodoRemove } from "../../actions/actionTodoRemove"
import { State } from "../../../types/State"
import { actionTodoUpdate } from "../../actions/actionTodoUpdate"

describe("reducerTodos", () => {
  it("handles actionTodoAdd", () => {
    // Arrange
    const stateInitial = undefined

    // Act
    const received = reducerTodos(
      stateInitial,
      actionTodoAdd({ id: "1", title: "Test Todo", complete: true }),
    )

    // Assert
    expect(received).toMatchSnapshot()
  })

  it("handles actionTodoRemove", () => {
    // Arrange
    const stateInitial: State["todos"] = {
      entities: {
        "1": { id: "1", title: "Test Todo", complete: true },
        "2": { id: "2", title: "Test Todo", complete: true },
      },
      result: ["1", "2"],
    }

    // Act
    const received = reducerTodos(stateInitial, actionTodoRemove("1"))

    // Assert
    expect(received).toMatchSnapshot()
  })

  it("handles actionTodoUpdate", () => {
    // Arrange
    const stateInitial: State["todos"] = {
      entities: {
        "1": { id: "1", title: "Test Todo", complete: true },
        "2": { id: "2", title: "Test Todo", complete: true },
      },
      result: ["1", "2"],
    }

    // Act
    const received = reducerTodos(
      stateInitial,
      actionTodoUpdate("2", { title: "I've changed" }),
    )

    // Assert
    expect(received).toMatchSnapshot()
  })
})
