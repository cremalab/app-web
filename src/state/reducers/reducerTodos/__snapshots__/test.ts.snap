// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducerTodos handles actionTodoAdd 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "complete": true,
      "id": "1",
      "title": "Test Todo",
    },
  },
  "result": Array [
    "1",
  ],
}
`;

exports[`reducerTodos handles actionTodoRemove 1`] = `
Object {
  "entities": Object {
    "2": Object {
      "complete": true,
      "id": "2",
      "title": "Test Todo",
    },
  },
  "result": Array [
    "2",
  ],
}
`;

exports[`reducerTodos handles actionTodoUpdate 1`] = `
Object {
  "entities": Object {
    "1": Object {
      "complete": true,
      "id": "1",
      "title": "Test Todo",
    },
    "2": Object {
      "complete": true,
      "id": "2",
      "title": "I've changed",
    },
  },
  "result": Array [
    "1",
    "2",
  ],
}
`;
